jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/ColorRefinementIsomorphismInspector.java

```diff
@@ -0,0 +1,299 @@
+ * as described in
+ * C. Berkholz, P. Bonsma, and M. Grohe.  Tight lower and upper bounds for the complexity of canonical
+ * colour refinement. Theory of Computing Systems,doi:10.1007/s00224-016-9686-0, 2016 (color refinement)
+ * The complexity of this algorithm is O(|V| + |E| log |V|).
+ *
+ * @param <V> the type of the vertices
+ * @param <E> the type of the edges
+ *
+ * @author Christoph Grüne
+ */
+public class ColorRefinementIsomorphismInspector<V, E> extends RefinementAbstractIsomorphismInspector<V, E> implements Serializable {
+
+    private static final long serialVersionUID = -4798546523147487442L;
+
+    /**
+     * The isomorphism that is calculated by this color refinement isomorphism inspector
+     */
+    private GraphMapping<V, E> isomorphicGraphMapping;
+
+    /**
+     * contains whether the graphs are isomorphic or not.
+     * If we cannot decide whether they are isomorphic the value will be not present.
+     */
+    private Boolean isIsomorphic;
+    /**
+     * contains whether the two graphs produce a discrete coloring.
+     * Then, we can decide whether the graphs are isomorphic.
+     */
+    private boolean isColoringDiscrete;
+    /**
+     * contains whether the two graphs are forests. Forests can be identified to be isomorphic or not.
+     */
+    private boolean isForest;
+
+    /**
+     * contains whether the isomorphism test is executed to ensure that every operation is defined all the time
+     */
+    private boolean isomorphismTestExecuted;
+
+    /**
+     * Constructor for a isomorphism inspector based on color refinement. It checks whether <code>graph1</code> and
+     * <code>graph2</code> are isomorphic.
+     *
+     * @param graph1 the first graph
+     * @param graph2 the second graph
+     */
+    public ColorRefinementIsomorphismInspector(Graph<V, E> graph1, Graph<V, E> graph2) {
+        super(graph1, graph2);
+        this.isomorphicGraphMapping = null;
+        this.isColoringDiscrete = false;
+        this.isomorphismTestExecuted = false;
+        this.isForest = false;
+    }
+
+    /**
+     * returns the mapping if the calculated coarse coloring returned by color refinement is discrete.
+     * Otherwise it returns null.
+     *
+     * @return the mapping of the graph isomorphism if the coloring is discrete, otherwise null.
+     */
+    @Override
+    public Iterator<GraphMapping<V, E>> getMappings() throws IllegalStateException {
+        if(!isomorphismTestExecuted) {
+            try {
+                isomorphismExistsWithUndecidableCase();
+            } catch(IsomorphismUndecidableException ignored) {
+
+            }
+        }
+        if(isIsomorphic != null && isIsomorphic) {
+            ArrayList<GraphMapping<V, E>> iteratorList = new ArrayList<>(1);
+            iteratorList.add(isomorphicGraphMapping);
+            return iteratorList.iterator();
+        } else {
+            return null;
```

Comment:
Better to return an iterator with no results. There are helper methods in `Collections`